agent/handlers/agentapi/taskprotection/v1/handlers/handlers.go

```diff
@@ -76,40 +76,41 @@ func UpdateTaskProtectionHandler(state dockerstate.TaskEngineState, credentialsM
 				loggerfield.Error: err,
 			})
 			writeJSONResponse(w, http.StatusBadRequest,
-				"Failed to decode request", updateTaskProtectionRequestType)
+				types.NewTaskProtectionResponseError(types.NewErrorResponsePtr("", ecs.ErrCodeInvalidParameterException,
+					"UpdateTaskProtection: failed to decode request"), nil),
+				updateTaskProtectionRequestType)
 			return
 		}
 
-		if request.ProtectionEnabled == nil {
-			writeJSONResponse(w, http.StatusBadRequest,
-				"Invalid request: does not contain 'ProtectionEnabled' field",
+		task, statusCode, errorCode, err := getTaskFromRequest(state, r)
+		if err != nil {
+			writeJSONResponse(w, statusCode,
+				types.NewTaskProtectionResponseError(types.NewErrorResponsePtr("", errorCode, err.Error()), nil),
 				updateTaskProtectionRequestType)
 			return
 		}
 
-		taskProtection, err := types.NewTaskProtection(*request.ProtectionEnabled, request.ExpiresInMinutes)
-		if err != nil {
+		if request.ProtectionEnabled == nil {
 			writeJSONResponse(w, http.StatusBadRequest,
-				fmt.Sprintf("Invalid request: %v", err),
+				types.NewTaskProtectionResponseError(types.NewErrorResponsePtr(task.Arn, ecs.ErrCodeInvalidParameterException,
+					"Invalid request: does not contain 'ProtectionEnabled' field"), nil),
 				updateTaskProtectionRequestType)
 			return
 		}
 
-		task, responseCode, err := getTaskFromRequest(state, r)
-		if err != nil {
-			writeJSONResponse(w, responseCode, err.Error(), updateTaskProtectionRequestType)
-			return
-		}
+		taskProtection := types.NewTaskProtection(*request.ProtectionEnabled, request.ExpiresInMinutes)
 
 		logger.Info("UpdateTaskProtection endpoint was called", logger.Fields{
 			loggerfield.Cluster:        cluster,
 			loggerfield.TaskARN:        task.Arn,
 			loggerfield.TaskProtection: taskProtection,
 		})
 
-		ecsClient, responseCode, err := factory.newTaskProtectionClient(credentialsManager, task)
+		ecsClient, statusCode, errorCode, err := factory.newTaskProtectionClient(credentialsManager, task)
```

Comment:
it feels a bit odd to have a client factory return http status and error code... Because creating the client is closer to interacting with upstream than validating donwstream request. 

Maybe the credential validation should be part of the downstream request validation (just like we check the request parameters) and we can throw the access denied exception. Only after all validation is done, we proceed to create the client to talk to ECS FE.

Reply:
Make sense to me. I will make this change.